train.py

```diff
@@ -0,0 +1,211 @@
+        model.eval()
+        # Evaluate the model on validation set
+        sloss, val_state = val_epoch(
+            (Batch(b[0], b[1], pad_idx) for b in valid_dataloader),
+            model,
+            SimpleLossCompute(module.generator, criterion),
+        )
+        # Log validation and training losses
+        print(sloss)
+        wandb.log({"val/loss": sloss, "train/loss": t_loss})
+
+    return model, run_id
+
+
+def train_epoch(
+    data_iter: Iterable,
+    model: nn.Module,
+    loss_compute: Callable,
+    optimizer: torch.optim.Optimizer,
+    scheduler: LambdaLR,
+    accum_iter=1,
+    train_state=TrainState(),
+) -> tuple[float, TrainState]:
+    start = time.time()
+    total_tokens = 0
+    total_loss = 0
+    tokens = 0
+    n_accum = 0
+    i = 0  # iteration counter
+
+    # create progress bar
+    pbar = tqdm(data_iter)
+    for batch in pbar:
+        out = model.forward(batch.src, batch.tgt, batch.src_mask, batch.tgt_mask)
+
+        loss, loss_node = loss_compute(out, batch.tgt_y, batch.ntokens)
+        loss_node.backward()
+
+        train_state.step += 1
+        train_state.samples += batch.src.shape[0]
+        train_state.tokens += batch.ntokens
+
+        # Update the model parameters and optimizer gradients every `accum_iter` iterations
+        if i % accum_iter == 0:
+            optimizer.step()
+            optimizer.zero_grad(set_to_none=True)
+            n_accum += 1
+            train_state.accum_step += 1
+        i += 1
+
+        # Update learning rate scheduler
+        scheduler.step()
+
+        # Update loss and token counts
+        total_loss += loss
+        total_tokens += batch.ntokens
+        tokens += batch.ntokens
+
+        # Display progress every 2 iterations
+        if i % 2 == 1:
+            lr = optimizer.param_groups[0]["lr"]
+            elapsed = time.time() - start
+            loss = loss / batch.ntokens
+            tok_rate = tokens / elapsed
+            pbar.set_description(
+                f"Epoch Step: {i:6d} | Accumulation Step: {n_accum:3d} | Loss: {loss:6.2f} | Tokens / Sec {tok_rate:7.1f} | Learning Rate: {lr:6.1e}"
+            )
+
+            # log the loss each to Weights and Biases
+            wandb.log({"train_steps/loss": loss})
+        del loss
+        del loss_node
+    # Return average loss over all tokens and updated train state
+    return total_loss / total_tokens, train_state
+
+
+def val_epoch(
+    data_iter: Iterable,
+    model: nn.Module,
+    loss_compute: Callable,
+    train_state=TrainState(),
+) -> tuple[float, TrainState]:
+    total_tokens = 0
+    total_loss = 0
+    tokens = 0
+    for i, batch in enumerate(data_iter):
+        out = model.forward(batch.src, batch.tgt, batch.src_mask, batch.tgt_mask)
+
+        loss, loss_node = loss_compute(out, batch.tgt_y, batch.ntokens)
+
+        total_loss += loss
+        total_tokens += batch.ntokens
+        tokens += batch.ntokens
+
+        del loss
+        del loss_node
```

Comment:
```suggestion
        total_loss += loss.item()
        total_tokens += batch.ntokens
        tokens += batch.ntokens
```

Not sure what's the point of using `del` here? If you were having problems with the loss being attached to the graph and your memory exploding, you could prevent that by using `.item()`. Same for the other epoch ⬆️